src/test/scala/mesosphere/marathon/api/v2/validation/PodsValidationTest.scala

```diff
@@ -124,6 +125,96 @@ class PodsValidationTest extends UnitTest with ValidationTestLike with PodsValid
+      val pod = validResidentPod.copy(scheduling = validResidentPod.scheduling.map(_.copy(
+        unreachableStrategy = Some(UnreachableEnabled()))))
+      validator(pod) should haveViolations(
+        "/" -> "unreachableStrategy must be disabled for pods with persistent volumes")
+    }
+
+    "be invalid if upgrade strategy has maximumOverCapacity set to non-zero" in new Fixture {
+      val pod = validResidentPod.copy(scheduling = validResidentPod.scheduling.map(_.copy(
+        upgrade = Some(PodUpgradeStrategy(maximumOverCapacity = 0.1)))))
+      validator(pod) should haveViolations(
+        "/upgrade/maximumOverCapacity" -> "got 0.1, expected 0.0")
+    }
+
+    "be invalid if cpu changes" in new Fixture {
+      val pod = validResidentPod.fromRaml
+      val to = pod.copy(containers = pod.containers.map(ct => ct.copy(resources = ct.resources.copy(cpus = 3))))
+      residentUpdateIsValid(pod)(to) should haveViolations(
+        "/" -> "cpus cannot be updated if a pod has persistent volumes")
+    }
+
+    "be invalid if mem changes" in new Fixture {
+      val pod = validResidentPod.fromRaml
+      val to = pod.copy(containers = pod.containers.map(ct => ct.copy(resources = ct.resources.copy(mem = 3))))
+      residentUpdateIsValid(pod)(to) should haveViolations(
+        "/" -> "mem cannot be updated if a pod has persistent volumes")
+    }
+
+    "be invalid if disk changes" in new Fixture {
+      val pod = validResidentPod.fromRaml
+      val to = pod.copy(containers = pod.containers.map(ct => ct.copy(resources = ct.resources.copy(disk = 3))))
+      residentUpdateIsValid(pod)(to) should haveViolations(
+        "/" -> "disk cannot be updated if a pod has persistent volumes")
+    }
+
+    "be invalid if gpus change" in new Fixture {
+      val pod = validResidentPod.fromRaml
+      val to = pod.copy(containers = pod.containers.map(ct => ct.copy(resources = ct.resources.copy(gpus = 3))))
+      residentUpdateIsValid(pod)(to) should haveViolations(
+        "/" -> "gpus cannot be updated if a pod has persistent volumes")
+    }
+
+    "be invalid with default upgrade strategy" in new Fixture {
+      val pod = validResidentPod.fromRaml
+      val to = pod.copy(upgradeStrategy = state.UpgradeStrategy.empty)
+      residentUpdateIsValid(pod)(to) should haveViolations(
+        "/upgradeStrategy/maximumOverCapacity" -> "got 1.0, expected 0.0")
+    }
+
+    "be invalid if persistent volumes change" in new Fixture {
+      val pod = validResidentPod.fromRaml
+      val to = pod.copy(volumes = pod.volumes.map {
+        case vol: PersistentVolume => vol.copy(persistent = vol.persistent.copy(size = 2))
+        case vol => vol
+      })
+      residentUpdateIsValid(pod)(to) should haveViolations(
+        "/" -> "persistent volumes cannot be updated")
+    }
+
+    "be invalid if ports change" in new Fixture {
+      val pod = validResidentPod.fromRaml
+      val to1 = pod.copy(containers = pod.containers.map(ct => ct.copy(
```

Comment:
Man, lens library would be nice for these.

Reply:
yeah, this is exactly the place where lens starts to make sense... :)

Reply:
but still these deep copies are a little bit of codesmell for me, because you might end up updating some deep internal part of the object and leave the object in state that does not make sense... but I guess it's ok-ish for tests :)